// ngx-page-scroll/esm2022/lib/ngx-page-scroll.module.mjs
import { NgModule } from '@angular/core';
import { NgxPageScrollCoreModule } from 'ngx-page-scroll-core';
import { NgxPageScrollDirective } from './ngx-page-scroll.directive';
import * as i0 from "@angular/core";
export class NgxPageScrollModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.6", ngImport: i0, type: NgxPageScrollModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.0.6", ngImport: i0, type: NgxPageScrollModule, declarations: [NgxPageScrollDirective], imports: [NgxPageScrollCoreModule], exports: [NgxPageScrollDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.0.6", ngImport: i0, type: NgxPageScrollModule, imports: [NgxPageScrollCoreModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.6", ngImport: i0, type: NgxPageScrollModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        NgxPageScrollDirective,
                    ],
                    imports: [
                        NgxPageScrollCoreModule,
                    ],
                    exports: [
                        NgxPageScrollDirective,
                    ],
                }]
        }] });
